---
title: "GEMMA phenotype input preparation, natural hybrids"
author: "Marta Binaghi"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document: 
    number_sections: yes
    toc: yes
editor_options: 
  chunk_output_type: inline
---
Created: Some time in June 2020
Last modified: 28/07/2022

# Preamble (setup and functions)

```{r setup, warning = FALSE}
knitr::opts_knit$set(root.dir = "/xxx/hybrids/")
knitr::opts_chunk$set(root.dir = "/xxx/hybrids/")
```

# Introduction

Observing distributions of phenotype data of the natural P. ax X P. ex hybrids.
Formatting the phenotype data for GWAS with gemma.
Observing correlations of phenotype with different variables.


# Phenotype descriptive and correlations

Phenotypes were loaded from the phenotype database that holds per plant mean of 
the phenotypes. The additional information on plant IDs is taken from the reads
sample ID file.


```{r pheno_import}
pheno_raw <- read.table("data/clean/phenotype_sequenced_individuals.csv",
                        header = TRUE,
                        sep = ",",
                        stringsAsFactors = FALSE)
# import plant IDs
labels <- read.table("data/clean/reads_sample_ID.csv",
                     sep = ",",
                     header = TRUE,
                     stringsAsFactors = FALSE)
# add a batch column
labels$batch <- 2018
labels$batch[grepl("KMH", labels$sra_library_id)] <- 2016
# remove extra lines that are repeated
labels <- labels[labels$lane == 1, ]
# labels have to be ordered as the beagle likelihood file.
labels <- labels[order(labels$id_gl), ]

pheno_raw$id_gl <- labels$id_gl[match(pheno_raw$new_id_short, labels$new_id_short)]
```

## Pistil - tube ratio

We calculate the ration between pistil and tube to represent the exertion of 
the pistil outside the tube.

```{r pheno_ratio}
pheno_raw$pist.tube.ratio <- pheno_raw$pist / pheno_raw$Dsum
```

## Normality

As tested in the phenotype analysis, the flavonol and anthocyanin are non normally distributed. They are thus rank-transformed.

```{r rank_pheno}
pheno_raw$flav_rank <- rank(pheno_raw$flav, na.last = "keep")
pheno_raw$antho_rank <- rank(pheno_raw$antho, na.last = "keep")
```

# Phenotype formatting for GEMMA

## Phenotype file format

Phenotype values were formatted and saved to be used in GEMMA. GEMMA handles
BIMBAM phenotype files, but can read single columns from a file with several
phenotype columns (space-separated). Phenotype has to be numeric, 
and each line is an individual
in the same order as in the genotype BIMBAM file (shown in the file 
`data/raw/gwas/hardfiltered_biallelic_cr09_mm005.bimbam.ids.txt`).

The resulting phenotype file `pheno_morpho.bimbam` includes in this order:

* Pistil-tube ratio
* Flavonols rank transformed
* Anthocyanins rank transformed


```{r pheno_format}
# read order of IDs
id_order <- read.table("data/raw/variants/hardfiltered_biallelic_cr09_mm005.bimbam.ids.txt",
                       header = FALSE)
# sort pheno_raw into id_order
pheno_sorted <- pheno_raw[match(id_order$V1, pheno_raw$new_id_short), ]
# get relevant columns (= phenotypes)
pheno_gwas <- as.data.frame(cbind(pheno_sorted$pist.tube.ratio))
colnames(pheno_gwas) <- c("Dsum", "pist", "pist.tube.ratio")
# rank transform
pheno_gwas$flav.rank <- rank(pheno_sorted$flav_rank, na.last = "keep")
pheno_gwas$antho.rank <- rank(pheno_sorted$antho_rank, na.last = "keep")
# save phenotype to file
write.table(pheno_gwas,
            file = "data/raw/gwas/pheno_gwas.bimbam",
            sep = " ",
            quote = FALSE,
            row.names = FALSE,
            col.names = FALSE)
```
